net/mwan3/files/usr/sbin/mwan3track

```diff
@@ -136,14 +143,23 @@ main() {
 			if [ $host_up_count -lt $reliability ]; then
 				case "$track_method" in
 					ping)
+						# Ensure that IPv6 address is reset
+						unset ADDR
 						if echo $track_ip | grep -q ':'; then
+							if "$KERNEL_HAS_SK_BUG"; then
+								# Pinging IPv6 hosts with an interface is troublesome on
+								# older kernel version. So get the IP address of the
+								# interface and use that instead.
+								ADDR=$(ip -6 addr ls dev "$DEVICE" | sed -ne '/\/128/d' -e '1,/ *inet6 /s/ *inet6 \([^ \/]*\).* scope global.*/\1/p')
+								[ -z "$ADDR" ] && ADDR=$(ip -6 addr ls dev "$DEVICE" | sed -ne 's/ *inet6 \([^ \/]*\).* scope global.*/\1/p')
+							fi
 							ping_protocol=6
 						fi
 						if [ $check_quality -eq 0 ]; then
-							$PING -$ping_protocol -I $DEVICE -c $count -W $timeout -s $size -t $max_ttl -q $track_ip &> /dev/null
+							$PING -$ping_protocol -I ${ADDR:-$DEVICE} -c $count -W $timeout -s $size -t $max_ttl -q $track_ip &> /dev/null
```

Comment:
I know you are trying to avoid another "if" and ash does not provide you with much resources. How about this instead of a string expansion?

```
set -- -$ping_protocol -c $count -W $timeout -s $size -t $max_ttl -q $track_ip
"$KERNEL_HAS_SK_BUG" && 
      set -- "$@" -I ADDR ||
      set -- "$@" -I $DEVICE
$PING "$@"
```

my 2 cents

Reply:
What don't you like about the code?
I think your proposed change a bit more confusing to me.
In the current code is more intuitive in my opinion and is native shell string manipulation.

Reply:
I agree with @feckert that the existing code is much (much) more readable than the suggestion.